src/main/java/org/jabref/logic/auxparser/DefaultAuxParser.java

```diff
@@ -146,26 +149,35 @@ private void resolveTags(AuxParserResult result) {
     /*
      * Resolves and adds CrossRef entries
      */
-    private void resolveCrossReferences(BibEntry entry, AuxParserResult result) {
-        entry.getField(StandardField.CROSSREF).ifPresent(crossref -> {
-            if (!result.getGeneratedBibDatabase().getEntryByKey(crossref).isPresent()) {
-                Optional<BibEntry> refEntry = masterDatabase.getEntryByKey(crossref);
-
-                if (refEntry.isPresent()) {
-                    insertEntry(refEntry.get(), result);
-                    result.increaseCrossRefEntriesCounter();
-                } else {
-                    result.getUnresolvedKeys().add(crossref);
+    private void resolveCrossReferences(List<BibEntry> entries, AuxParserResult result) {
+        List<BibEntry> entriesToInsert = new ArrayList<>();
+        for (BibEntry entry : entries) {
+            entry.getField(StandardField.CROSSREF).ifPresent(crossref -> {
+                if (!result.getGeneratedBibDatabase().getEntryByKey(crossref).isPresent()) {
+                    Optional<BibEntry> refEntry = masterDatabase.getEntryByKey(crossref);
+
+                    if (refEntry.isPresent()) {
+                        if (!entriesToInsert.contains(refEntry.get())) {
+                            entriesToInsert.add(refEntry.get());
+                            result.increaseCrossRefEntriesCounter();
+                        }
+                    } else {
+                        result.getUnresolvedKeys().add(crossref);
+                    }
                 }
-            }
-        });
+            });
+        }
+        insertEntries(entriesToInsert, result);
     }
 
     /*
-     * Insert a clone of the given entry. The clone is given a new unique ID.
+     * Insert a clone of each given entry. The clones are each given a new unique ID.
      */
-    private void insertEntry(BibEntry entry, AuxParserResult result) {
-        BibEntry clonedEntry = (BibEntry) entry.clone();
-        result.getGeneratedBibDatabase().insertEntry(clonedEntry);
+    private void insertEntries(List<BibEntry> entries, AuxParserResult result) {
+        List<BibEntry> clonedEntries = new ArrayList<>();
```

Comment:
Can we do following code here?

```java
List<BibEntry> clonedEntries = entries.stream().map(BibEntry::clone).collect(Collectors.toList())
```

Reply:
I don't know if you have to cast the `BibEntry`s from `Object`s to `BibEntry`s. I could try it out, though.

Reply:
The compiler complains

```
/Users/Abe/Java/jabref/src/main/java/org/jabref/logic/auxparser/DefaultAuxParser.java:186: error: incompatible types: inference variable T has incompatible bounds
        List<BibEntry> clonedEntries = entries.stream().map(BibEntry::clone).collect(Collectors.toList());
                                                                                    ^
    equality constraints: BibEntry
    lower bounds: Object
  where T is a type-variable:
    T extends Object declared in method <T>toList()
```

Reply:
I don´t think the clone is actually necessary, can you please try to remove it. Thanks